intervals/interval.py

```diff
@@ -667,11 +667,45 @@ def is_connected(self, other):
           [3, 3)
         * [1, 3) and (3, 5) are not connected
         """
-        return self.upper > other.lower and other.upper > self.lower or (
-            self.upper == other.lower and (self.upper_inc or other.lower_inc)
-        ) or (
-            self.lower == other.upper and (self.lower_inc or other.upper_inc)
-        )
+        gap = self.gap_interval(other)
+        return gap is None or gap.empty
+
+    def gap_interval(self, other):
+        """Find the gap between the two given intervals if it exists.
+
+        Examples:
+
+        * [1, 3] and [5, 7] have a gap of (3, 5)
+        * [5, 7] and [1, 3] have a gap of (3, 5)
+        * [2, 4) and [3, 5) do not have a gap
+        * [1, 3) and [3, 5) have an empty gap of [3, 3)
+        * [1, 3] and [4, 5] have an empty gap of (3, 4)
+        * [1, 3) and (3, 5) have a gap of [3, 3]
+        """
+        lower = self.upper
+        lower_inc = self.upper_inc
+        if other.upper < self.upper:
+            lower = other.upper
+            lower_inc = other.upper_inc
+        elif self.upper == other.upper:
```

Comment:
I'd swap the order of this `elif `comparison operands so that it's immediately differentiable from the `if` statement:

i.e.

```python
if other.upper < self.upper:
...
elif other.upper == self.upper:
...
```